### ISSUE TITLE
calamari-cross-fold-train gets stuck on broken PNG file

### ISSUE THREAD

**Issue by mikegerber** [Calamari-OCR/calamari#124]
I've been trying to train Calamari on the GT4HistOCR dataset. This dataset contains one broken PNG file (`1882-keller_sinngedicht/04970.nrm.png` in `corpus/dta19.tar.bz2`). When pointing calamari-cross-fold-train on this, the training gets stuck:

~~~
$ calamari-cross-fold-train --files BAD-1882-keller_sinngedicht/*.png --best_models_dir models --early_stopping_nbest 10 --max_parallel_models 2
Resolving input files
Found 100 files in the dataset
Executing: /srv/tmp/2019-07-18-trigger-calamari-train-bug/venv/bin/python3 -u /srv/tmp/2019-07-18-trigger-calamari-train-bug/venv/lib/python3.7/site-packages/calamari_ocr/scripts/train.py --files /tmp/calamarizt0l9vh2/fold_0.json
Executing: /srv/tmp/2019-07-18-trigger-calamari-train-bug/venv/bin/python3 -u /srv/tmp/2019-07-18-trigger-calamari-train-bug/venv/lib/python3.7/site-packages/calamari_ocr/scripts/train.py --files /tmp/calamarizt0l9vh2/fold_1.json
FOLD 0 | Resolving input files
FOLD 0 | Found 80 files in the dataset
FOLD 0 | Resolving validation files
FOLD 0 | Found 20 files in the validation dataset
FOLD 1 | Resolving input files
FOLD 1 | Found 80 files in the dataset
FOLD 1 | Resolving validation files
FOLD 1 | Found 20 files in the validation dataset
FOLD 0 | Preloading dataset type DataSetMode.TRAIN with size 80
Preloading data:   0%|                                                                                                                 | 0/80 [00:00<?, ?it/s]FOLD 1 | Preloading dataset type DataSetMode.TRAIN with size 80
Preloading data: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████| 80/80 [00:01<00:00, 58.42it/s]
FOLD 0 | Preloading dataset type DataSetMode.TRAIN with size 20
Preloading data:   0%|                                                                                                                 | 0/20 [00:00<?, ?it/s]Traceback (most recent call last):
  File "/usr/lib64/python3.7/multiprocessing/queues.py", line 236, in _feed
    obj = _ForkingPickler.dumps(obj)
  File "/usr/lib64/python3.7/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
  File "/srv/tmp/2019-07-18-trigger-calamari-train-bug/venv/lib/python3.7/site-packages/PIL/Image.py", line 724, in __getstate__
    return [self.info, self.mode, self.size, self.getpalette(), self.tobytes()]
  File "/srv/tmp/2019-07-18-trigger-calamari-train-bug/venv/lib/python3.7/site-packages/PIL/Image.py", line 1355, in getpalette
    self.load()
  File "/srv/tmp/2019-07-18-trigger-calamari-train-bug/venv/lib/python3.7/site-packages/PIL/ImageFile.py", line 234, in load
    s = read(self.decodermaxblock)
  File "/srv/tmp/2019-07-18-trigger-calamari-train-bug/venv/lib/python3.7/site-packages/PIL/PngImagePlugin.py", line 658, in load_read
    cid, pos, length = self.png.read()
  File "/srv/tmp/2019-07-18-trigger-calamari-train-bug/venv/lib/python3.7/site-packages/PIL/PngImagePlugin.py", line 126, in read
    raise SyntaxError("broken PNG file (chunk %s)" % repr(cid))
  File "<string>", line None
SyntaxError: broken PNG file (chunk b'\xd5\xf7\xbf3')
Traceback (most recent call last):
  File "/usr/lib64/python3.7/multiprocessing/queues.py", line 236, in _feed
    obj = _ForkingPickler.dumps(obj)
  File "/usr/lib64/python3.7/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
  File "/srv/tmp/2019-07-18-trigger-calamari-train-bug/venv/lib/python3.7/site-packages/PIL/Image.py", line 724, in __getstate__
    return [self.info, self.mode, self.size, self.getpalette(), self.tobytes()]
  File "/srv/tmp/2019-07-18-trigger-calamari-train-bug/venv/lib/python3.7/site-packages/PIL/Image.py", line 1355, in getpalette
    self.load()
  File "/srv/tmp/2019-07-18-trigger-calamari-train-bug/venv/lib/python3.7/site-packages/PIL/ImageFile.py", line 234, in load
    s = read(self.decodermaxblock)
  File "/srv/tmp/2019-07-18-trigger-calamari-train-bug/venv/lib/python3.7/site-packages/PIL/PngImagePlugin.py", line 658, in load_read
    cid, pos, length = self.png.read()
  File "/srv/tmp/2019-07-18-trigger-calamari-train-bug/venv/lib/python3.7/site-packages/PIL/PngImagePlugin.py", line 126, in read
    raise SyntaxError("broken PNG file (chunk %s)" % repr(cid))
  File "<string>", line None
SyntaxError: broken PNG file (chunk b'\xd5\xf7\xbf3')
Preloading data:  20%|█████████████████████                                                                                    | 4/20 [00:00<00:00, 37.48it/s]
~~~

At this point, training is stuck – no more output, no more progress. If I remove the bad file, everything works. Tested using calamari-ocr 0.3.5.

I've attached the test files: [BAD-GOOD-1882-keller_sinngedicht.tar.gz](https://github.com/Calamari-OCR/calamari/files/3406733/BAD-GOOD-1882-keller_sinngedicht.tar.gz)

**Comment by ChWick** [comment]
Damn it, I forgot about your issue and ran into the same bug. I resolved this issue with the latest commit c4d8fdd5f2c849a6d316479caee79d136ebc0f5a. The bug was caused by PIL loading images lazy, thus the exception was not thrown where I expected it...

**Comment by mikegerber** [comment]
Thanks :)
